src/appleseed/renderer/kernel/shading/closures.h

```diff
@@ -117,10 +122,54 @@ enum ClosureID
 struct ExceptionOSLRuntimeError
   : public foundation::Exception
 {
-    explicit ExceptionOSLRuntimeError(const char* what)
-      : foundation::Exception(what)
+    explicit ExceptionOSLRuntimeError(const char* what);
+};
+
+
+//
+// Small closure layer ID stack.
+//
+
+template <size_t N>
+class ClosureLayerIDStack
+{
+  public:
+    ClosureLayerIDStack()
+      : m_size(0)
     {
     }
+
+    size_t size() const
+    {
+        return m_size;
+    }
+
+    bool empty() const
+    {
+        return m_size == 0;
+    }
+
+    std::int8_t operator[](const size_t i) const
+    {
+        assert(i < m_size);
+        return m_ids[i];
+    }
+
+    void push(const std::int8_t id)
+    {
+        assert(m_size < N - 1);
+        m_ids[m_size++] = id;
+    }
+
+    void pop()
+    {
+        assert(m_size > 0);
+        --m_size;
+    }
+
+  private:
+    std::int8_t   m_ids[N];
```

Comment:
Shouldn't we use `std::uint8_t` instead? Or do we need to allow negative IDs?

Reply:
negative numbers are used to signal that there are no layers.